mindmeld/components/nlp.py

```diff
@@ -610,6 +610,30 @@ def process_query(
             processed_query.confidence = scores
         return processed_query
 
+    def _update_nlp_hierarchy(self, nlp_components, domain, intent, entity=None, role=None):
+        # We assume that the intent is a correct child of the domain
+        if domain not in nlp_components:
+            nlp_components[domain] = {}
+
+        if intent not in nlp_components[domain]:
+            nlp_components[domain][intent] = {}
+
+        all_entities_intent = self.domains[domain].intents[intent].entities
+        valid_entities = filter(lambda candidate: entity and entity in {'*', candidate},
+                                all_entities_intent)
+
+        for nlp_entity in valid_entities:
+            if nlp_entity not in nlp_components[domain][intent]:
+                nlp_components[domain][intent][nlp_entity] = {}
+
+            all_roles_in_entity = self.domains[
+                domain].intents[intent].entities[nlp_entity].role_classifier.roles
+            valid_roles = filter(lambda candidate: role and role in {'*', candidate},
+                                 all_roles_in_entity)
+
+            for nlp_role in valid_roles:
+                nlp_components[domain][intent][nlp_entity][nlp_role] = {}
+
     def extract_allowed_intents(self, allowed_intents):
```

Comment:
Nit: Maybe we should replace the `allowed_intents` naming to `allowed_nlp_classes`. If that's too much hassle, we should update the comment to reflect this new functionality.

Reply:
done